### PR TITLE
[xbmc] move Environment to platform and split to platform specific implementations.

### PR DESCRIPTION
**Pull request by afedchin**
see title

### CODE REVIEW

File: xbmc/platform/win10/Environment.cpp [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+using namespace Windows::Storage;
+using namespace Windows::Foundation;
+
+// --------------------- Internal Function ---------------------
+
+/**
+ * \fn int CEnvironment::win_setenv(const std::wstring &name, const std::wstring &value = L"",
+ *     updateAction action = autoDetect)
+ * \brief Internal function used to manipulate with environment variables on win32.
```
Comment (notspiff) [inline xbmc/platform/win10/Environment.cpp]: s/with//

File: xbmc/platform/win32/Environment.cpp [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+ * 		   several errors, sum of all errors values; non-zero in case of other errors.
+ */
+int CEnvironment::win_setenv(const std::string &name, const std::string &value /* = "" */, enum updateAction action /* = autoDetect */)
+{
+  std::wstring Wname = KODI::PLATFORM::WINDOWS::ToW(name);
+  if (Wname.empty() || name.find('=') != std::wstring::npos)
+    return -1;
+  if ( (action == addOnly || action == addOrUpdateOnly) && value.empty() )
+    return -1;
+  if (action == addOnly && !(getenv(name).empty()) )
```
Comment (notspiff) [inline xbmc/platform/win32/Environment.cpp]: unnecessary bracing

File: xbmc/platform/win32/Environment.cpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+#ifdef _DEBUG
+    { L"msvcr120d.dll" },// Visual Studio 2013 (debug)
+#endif
+    { L"vcruntime140.dll" },
+    { L"ucrtbase.dll" },
+#ifdef _DEBUG
+    { L"vcruntime140d.dll" },
+    { L"ucrtbased.dll" },
+#endif
+    { NULL }             // Terminating NULL for list
```
Comment (notspiff) [inline xbmc/platform/win32/Environment.cpp]: nullptr

File: xbmc/platform/win32/Environment.cpp [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+  };
+
+  // Check all modules each function run, because modules can be loaded/unloaded at runtime
+  for (int i = 0; modulesList[i]; i++)
+  {
+    HMODULE hModule;
+    if (!GetModuleHandleExW(0, modulesList[i], &hModule) || hModule == NULL) // Flag 0 ensures that module will be kept loaded until it'll be freed
```
Comment (notspiff) [inline xbmc/platform/win32/Environment.cpp]: nullptr. and more in this function.

File: xbmc/platform/win10/Environment.cpp [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+ * \return Zero on success and -1 otherwise
+ */
+int CEnvironment::win_setenv(const std::string &name, const std::string &value /* = "" */, enum updateAction action /* = autoDetect */)
+{
+  std::wstring Wname = KODI::PLATFORM::WINDOWS::ToW(name);
+  if (Wname.empty() || name.find('=') != std::wstring::npos)
+    return -1;
+  if ((action == addOnly || action == addOrUpdateOnly) && value.empty())
+    return -1;
+  if (action == addOnly && !(getenv(name).empty()))
```
Comment (notspiff) [inline xbmc/platform/win10/Environment.cpp]: unnecessary bracing.

**Review by notspiff — APPROVED** [review]
some minor suggestion that i spotted. looks good, much better than the spaghetti

**Comment by FernetMenta** [comment]
did you use git move?

**Comment by afedchin** [comment]
yes, see the first commit. I don't know why github shows it different

**Comment by afedchin** [comment]
`Android-ARM` failure is unrelated

File: xbmc/platform/win32/Environment.cpp [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  std::wstring Wvalue = KODI::PLATFORM::WINDOWS::ToW(value);
+  int retValue = 0;
+
+  std::wstring EnvString;
+  if (action == deleteVariable)
+    EnvString = Wname + L"=";
+  else
+    EnvString = Wname + L"=" + Wvalue;
+
+  static const wchar_t *modulesList[] =
```
Comment (Paxxi) [inline xbmc/platform/win32/Environment.cpp]: we can get rid of this. It's not an issue anymore after we unified our crt depends.

**Review by Paxxi — COMMENTED** [review]
We can get rid of the module loading, env injections tricks and just call putenv directly.

It can be done in a separate PR if there's something depending on this getting merged.

File: xbmc/platform/win32/Environment.cpp (hunk 6 above)
Comment (afedchin) [reply xbmc/platform/win32/Environment.cpp]: just to be clarify. does not it mater even if we update CRT to v141?

**Review by Paxxi — APPROVED** [review]

**Pull request merged**